### ISSUE TITLE
Stopped working after upgrade to UniFi OS 1.12.38

### ISSUE THREAD

**Issue by snoack** [peacey/split-vpn#160]
After my Dream Machine Pro updated to UniFi OS 1.12.38 last night, the VPN stopped working. It works with the same OpenVPN configuration from my laptop, and it did work yesterday on the UDM-Pro before the update. Now the network with the VPN remains offline, and I get following error in the OpenVPN log file: `[UNDEF] Inactivity timeout (--ping-restart), restarting`

**Comment by peacey** [comment]
Hello @snoack,

Can you show me the full OpenVPN log?

**Comment by peacey** [comment]
Weirdly, it just looks like the UDM isn't able to connect to it. split-vpn doesn't even have a chance to run since the connection never comes up. 

Do you have any firewall rules that could affect this? Is it possible to try a different server?

Am I able to test this server? Is it a public one?

**Comment by snoack** [comment]
Yeah, feel free to test with my OpenVpn config. It's using Riseup's US east coast VPN servers, they are public. I put links to download the CA and certificate below the config in my previous comment (or you can use my [bitmask-openvpn](https://github.com/snoack/bitmask-openvpn) script to generate the config and download the certificate).

Unfortunately, I don't have another VPN server to test it with. If you do, I'm happy to test it with your server.

**Comment by peacey** [comment]
Okay, so I think I figured out the issue. I tested your server and had the same problem on the UDMP, but it worked on my other system (like you). Other servers I tried on the UDMP didn't have any issues, so it was isolated to your server. After turning on debug flags on OpenVPN (`--verb 4`), I saw that it was stalling at the initial TLS handshake (and I could see the server reply in tcpdump so connection was not being blocked). Then I realized that the UDM's openvpn is using an old version of the openssl library from 2019 (1.0.2u) by checking `openvpn --version`, while my other system was using a much newer one (> 3.0). I cross-compiled a static openvpn binary for the UDM with the newest openssl (3.0.8) and tested it on the UDM and it worked with your server! Probably openssl 1.1 or 1.2 would have worked fine too, though I didn't check what's the minimum working version.

So that leads me to think that either your server does not support older openssl libraries, or Ubiquiti did something weird with their openssl library. It's difficult to tell what went wrong without the server logs, so I'm not really sure why it was working before. I also can't really check what the openssl library version was on previous firmwares to confirm my theory (I'm on 2.4 so I can't downgrade to 1.x firmwares anymore, but on 2.4 the openssl library is 1.0.2u and you probably have the same on 1.12.38).

In any case, you can download my statically compiled openvpn binary (or statically compile it yourself), place it in your UDM's riseup vpn folder, and use it instead of the system openvpn. Here's how:

  ```sh
  cd /etc/split-vpn/openvpn/riseup
  curl -LO https://github.com/peacey/split-vpn/raw/static_openvpn/bin/openvpn
  chmod +x ./openvpn
  ./openvpn --version
  ```

  * Make sure to change the first cd line to your correct directory, and check that the last command shows the new openvpn/openssl version and doesn't error.

Then, edit your `run-vpn.sh` in the same folder to use `./openvpn` from this folder instead of the system `openvpn` by changing the line `nohup openvpn ...` to `nohup ./openvpn ...`.

See if that works for you. 

Also, if your VPN provider is nice and you're curious why it's not working with the old openssl version, I would probably ask them if they can verify what's the minimum TLS/openssl version they support or check their server logs for what was going wrong.

**Comment by snoack** [comment]
Apparently as of UniFi OS 1.12.38, OpenVPN is using OpenSSL 1.1.1n:

```
OpenVPN 2.5.2 aarch64-buildroot-linux-gnu [SSL (OpenSSL)] [LZO] [LZ4] [EPOLL] [MH/PKTINFO] [AEAD] built on Jan 19 2023
library versions: OpenSSL 1.1.1n  15 Mar 2022, LZO 2.10
Originally developed by James Yonan
Copyright (C) 2002-2018 OpenVPN Inc <sales@openvpn.net>
Compile time defines: enable_async_push=no enable_comp_stub=no enable_crypto_ofb_cfb=yes enable_debug=yes enable_def_auth=yes enable_dependency_tracking=no enable_dlopen=unknown enable_dlopen_self=unknown enable_dlopen_self_static=unknown enable_doc=no enable_docs=no enable_documentation=no enable_fast_install=yes enable_fragment=yes enable_gtk_doc=no enable_gtk_doc_html=no enable_iproute2=no enable_ipv6=yes enable_libtool_lock=yes enable_lz4=yes enable_lzo=yes enable_management=yes enable_multihome=yes enable_nls=no enable_pam_dlopen=no enable_pedantic=no enable_pf=yes enable_pkcs11=no enable_plugin_auth_pam=yes enable_plugin_down_root=yes enable_plugins=yes enable_port_share=yes enable_selinux=no enable_shared=yes enable_shared_with_static_runtimes=no enable_small=no enable_static=no enable_strict=no enable_strict_options=no enable_systemd=no enable_unit_tests=no enable_werror=no enable_win32_dll=yes enable_x509_alt_username=no with_aix_soname=aix with_crypto_library=openssl with_fop=no with_gnu_ld=yes with_mem_check=no with_sysroot=no with_xmlto=no
```

Also I figured out that if I strip the line `tls-cipher DHE-RSA-AES128-SHA` from the config it does work fine. I guess that's what I gonna do. Thanks for your help, it wouldn't have ooccured to me without your pointer toward OpenSSL.

> Also, if your VPN provider is nice and you're curious why it's not working with the old openssl version, I would probably ask them if they can verify what's the minimum TLS/openssl version they support or check their server logs for what was going wrong.

Well, officially Riseup only supports their [Bitmask](https://github.com/leapcode/bitmask-vpn)  (desktop) client. [bitmask-openvpn](https://github.com/snoack/bitmask-openvpn) is an effort by myself to reverse engineer their client specifically to use Riseup VPN with plain OpenVPN and split-vpn on the network level.

**Comment by peacey** [comment]
Oh damn, I should've tried to change the cipher first lol. I'm surprised it did work for me with the updated openssl library though with the same cipher you initially used, so clearly there is some incompatibility with that cipher and the current openssl on your system and mine.

Well anyway, I'm glad you figured it out and found an easier solution than mine!
